sbt-scalajs-bundler/src/main/scala/scalajsbundler/PackageManager.scala

```diff
@@ -0,0 +1,252 @@
+package scalajsbundler
+
+import java.io.File
+
+import sbt._
+import scalajsbundler.util.Commands
+import scalajsbundler.util.JSON
+
+trait PackageManager {
+
+  val name: String
+
+  /**
+    * Runs the command `cmd`
+    * @param args Command arguments
+    * @param workingDir Working directory of the process
+    * @param logger Logger
+    */
+  def run(args: String*)(workingDir: File, logger: Logger): Unit
+
+  def install(baseDir: File, installDir: File, logger: Logger): Unit
+
+  val packageJsonContents: Map[String, JSON]
+}
+
+object PackageManager {
+
+  abstract class ExternalProcess(
+    val name: String,
+    val installCommand: String,
+    val installArgs: Seq[String]
+  ) extends PackageManager {
+
+    def run(args: String*)(workingDir: File, logger: Logger): Unit =
+      Commands.run(cmd ++: args, workingDir, logger)
+
+    private val cmd = sys.props("os.name").toLowerCase match {
+      case os if os.contains("win") => Seq("cmd", "/c", name)
+      case _                        => Seq(name)
+    }
+
+    def install(baseDir: File, installDir: File, logger: Logger): Unit = {
+      this match {
+        case lfs: LockFileSupport =>
+          lfs.lockFileRead(baseDir, installDir, logger)
+        case _ =>
+          ()
+      }
+
+      run(installCommand +: installArgs: _*)(installDir, logger)
+
+      this match {
+        case lfs: LockFileSupport =>
+          lfs.lockFileWrite(baseDir, installDir, logger)
+        case _ =>
+          ()
+      }
+    }
+  }
+
+  trait AddPackagesSupport { this: PackageManager =>
+
+    val addPackagesCommand: String
+    val addPackagesArgs: Seq[String]
+
+    /**
+      * Locally install NPM packages
+      *
+      * @param baseDir The (sub-)project directory which contains yarn.lock
+      * @param installDir The directory in which to install the packages
+      * @param logger sbt logger
+      * @param npmPackages Packages to install (e.g. "webpack", "webpack@2.2.1")
+      */
+    def addPackages(baseDir: File,
+                    installDir: File,
+                    logger: Logger,
+                   )(npmPackages: String*): Unit = {
+      this match {
+        case lfs: LockFileSupport  =>
+          lfs.lockFileRead(baseDir, installDir, logger)
+        case _ =>
+          ()
+      }
+
+      run(addPackagesCommand +: (addPackagesArgs ++ npmPackages): _*)(installDir, logger)
+
+      this match {
+        case lfs: LockFileSupport  =>
+          lfs.lockFileWrite(baseDir, installDir, logger)
+        case _ =>
+          ()
+      }
+    }
+  }
+
+  trait LockFileSupport {
+    val lockFileName: String
```

Comment:
It would be better to convert this to a `File` to allow for better lockfile location customization.

Reply:
Though I do not have the best idea on where to put `PackageManager.Npm` and `PackageManager.Yarn` so that `baseDirectory` is available during construction.